Logging: add an option to turn logging off altogether

josephsl opened the pull request:
### Link to issue number:
Fixes #8516

### Summary of the issue:
Add an option to turn off logging altogether.

### Description of how this pull request fixes the issue:
Add a new option in General settings panel to turn off logging. If this option is chosen, only a startup message (along with config path) will be logged, and log level will be set to 100.

### Testing performed:
Tested on source code with logging level set to off and other values.

### Known issues with pull request:
Even with logging level set to off, startup info messages will be shown up to config path. Ideally this too should not be logged, but for now, it is acceptable.

### Change log entry:
New features: It is possible to turn off NVDA's logging features off by setting logging level to "disabled" from General settings panel. (#8516)

Thanks.

josephsl committed:
Log handler: update copyright headers

josephsl committed:
Log handler/level names: Python 3 split this into two dictionaries (note).

josephsl committed:
Add a new command-line switch (--no-logging) that'll completely turn off logging functionality. re #8516.

If NVDA is run with --no-logging, logging facility will be turned off altogether. This is handy for not only testing this feature, but also for those who'd like to add this command-line switch while defining nVDA shortcut on the desktop or add this from Run dialog (Windows+R).

josephsl committed:
Log handler: add support for no logging switch. Re #8516.

Switch to null logger if no logging flag is turned on via command-line switches. Also, detect log level of 'OFF' and set log level accordingly (to 100).

josephsl committed:
Log handler: log level 100 -> Logger.OFF. Re #8516.

josephsl committed:
Settings/General panel: add 'disabled' as a log level. Re #8516.

josephsl committed:
User guide: document log levels and improved command-line description for secure mode. Re #8516.

Describe available logging modes and what they mean. Also, when edscribing secure mode, add the fact that logging is turned off. Lastly, add an entry in command-line switches table for --no-logging switch.

josephsl committed:
Merge branch 'master' into i8516noLogging

leonardder commented:
Could you please use the closes or fixes keyword for the Link to issue number? In that case, github magic will automatically close the issue when the pr is merged.

josephsl commented:
Hi, I see – we may need to document this in “contributing” guide. Thanks for the reminder.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Sunday, August 5, 2018 11:40 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Logging: add an option to turn logging off altogether (#8596)

 

Could you please use the closes or fixes keyword for the Link to issue number? In that case, github magic will automatically close the issue when the pr is merged.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8596#issuecomment-410604147> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkAjlVznU1Cg-k5ypCxO1SsJsI23Sks5uN-SogaJpZM4Vv0U1> .

Brian1Gaff commented:
Not quite sure I agree with the premise behind this. Is a log where it is 
any less secure than other data in similar places on the users machine??

Its no good plugging a hole in a boat with no bottom.
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Joseph Lee" <notifications@github.com>
To: "nvaccess/nvda" <nvda@noreply.github.com>
Cc: "Subscribed" <subscribed@noreply.github.com>
Sent: Monday, August 06, 2018 7:23 AM
Subject: [nvaccess/nvda] Logging: add an option to turn logging off 
altogether (#8596)


> ### Link to issue number:
> #8516
>
> ### Summary of the issue:
> Add an option to turn off logging altogether.
>
> ### Description of how this pull request fixes the issue:
> Add a new option in General settings panel to turn off logging. If this 
> option is chosen, only a startup message (along with config path) will be 
> logged, and log level will be set to 100.
>
> ### Testing performed:
> Tested on source code with logging level set to off and other values.
>
> ### Known issues with pull request:
> Even with logging level set to off, startup info messages will be shown up 
> to config path. Ideally this too should not be logged, but for now, it is 
> acceptable.
>
> ### Change log entry:
> New features: It is possible to turn off NVDA's logging features off by 
> setting logging level to "disabled" from General settings panel. (#8516)
>
> Thanks.
> You can view, comment on, or merge this pull request online at:
>
>  https://github.com/nvaccess/nvda/pull/8596
>
> -- Commit Summary --
>
>  * Log handler: update copyright headers
>  * Log handler/level names: Python 3 split this into two dictionaries 
> (note).
>  * Add a new command-line switch (--no-logging) that'll completely turn 
> off logging functionality. re #8516.
>  * Log handler: add support for no logging switch. Re #8516.
>  * Log handler: log level 100 -> Logger.OFF. Re #8516.
>  * Settings/General panel: add 'disabled' as a log level. Re #8516.
>  * User guide: document log levels and improved command-line description 
> for secure mode. Re #8516.
>  * Merge branch 'master' into i8516noLogging
>
> -- File Changes --
>
>    M source/gui/settingsDialogs.py (2)
>    M source/logHandler.py (16)
>    M source/nvda.pyw (15)
>    M user_docs/en/userGuide.t2t (17)
>
> -- Patch Links --
>
> https://github.com/nvaccess/nvda/pull/8596.patch
> https://github.com/nvaccess/nvda/pull/8596.diff
>
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/8596
>

DrSooom commented:
Shall the default value set to "Info" (actual behaviour in NVDA Alpha 15781, 2018-08-03; General NVDA Settings window) or to "Disabled" (privacy by default)?

Furthermore there are two different nomenclature for the log levels in the User Guide.
> Disabled, Info, Debug warning, Input/output Debug
> | -l LOGLEVEL | --log-level=LOGLEVEL | The lowest level of message logged (debug 10, info 20, warning 30, error 40, critical 50, disabled 100), default is warning |

I'm a little bit confused.

josephsl commented:
Hi, this code hasn’t been merged into master (alpha) branch yet because it needs to go through reviews. The two sets of log levels can be explained by the fact that one is user-friendly one for settings whereas the other one is for command-line use. Thanks.

 

From: Daniel Mayr <notifications@github.com> 
Sent: Tuesday, August 7, 2018 2:38 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Logging: add an option to turn logging off altogether (#8596)

 

Shall the default value set to "Info" (actual behaviour in NVDA Alpha 15781, 2018-08-03; General NVDA Settings window) or to "Disabled" (privacy by default)?

Furthermore there are two different nomenclature for the log levels in the User Guide.

Disabled, Info, Debug warning, Input/output Debug
| -l LOGLEVEL | --log-level=LOGLEVEL | The lowest level of message logged (debug 10, info 20, warning 30, error 40, critical 50, disabled 100), default is warning |

I'm a little bit confused.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8596#issuecomment-410997209> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkAeZOEB2CQIaDa7-dnEEtYjiWc5Wks5uOV_ggaJpZM4Vv0U1> .

DrSooom commented:
I looked at the changed files in this PR directly via the GitHub website.

After merging there will be five log levels to be set up via the NVDA GUI and six via the command line. It would be nice to know which log level number have the five options which can be set via the GUI. Or as a question: What log level I have to choose in the command line to get the same result like with "Input/Output" via the GUI?

josephsl committed:
Merge branch 'master' into i8516noLogging

josephsl committed:
User guide: add log levels for I/O and debugwarning log levels.

derekriemer reviewed (commented):
good. a few minor thoughts.

josephsl committed:
Merge branch 'master' into i8516noLogging

josephsl commented:
Hi,

Although I have no comments about nvda.pyw, I'll edit the user guide based on feedback. Thanks.

josephsl committed:
User guide: address review actions.

Add notes about privacy for I/O and debug logging levels, along with word usage changes.

josephsl committed:
Merge branch 'master' into i8516noLogging

josephsl commented:
Hi,

@feerrenrut, if you don't mind, I'd like to get a second look please. Thanks.

josephsl committed:
Merge branch 'master' into i8516noLogging

feerrenrut commented on source/nvda.pyw:
```diff
@@ -189,12 +190,16 @@ if isSecureDesktop:
-if globalVars.appArgs.debugLogging:
-	logLevel=log.DEBUG
+if globalVars.appArgs.noLogging:
+	logLevel = 100
```
this could use the `log.OFF` constant, right? EG:
```suggestion
	logLevel = log.OFF
```

feerrenrut commented on source/logHandler.py:
```diff
@@ -350,14 +355,15 @@ def initialize(shouldDoRemoteLogging=False):
-	if not level or level > log.INFO:
+	# The lone exception to level higher than INFO is "OFF" (100).
+	if not level or log.INFO < level < log.OFF:
 		log.warning("invalid setting for logging level: %s" % levelName)
```
Interestingly, I suspect this warning could never end up in the log.

feerrenrut commented on source/logHandler.py:
```diff
@@ -350,14 +355,15 @@ def initialize(shouldDoRemoteLogging=False):
 	levelName=config.conf["general"]["loggingLevel"]
 	level = levelNames.get(levelName)
-	if not level or level > log.INFO:
+	# The lone exception to level higher than INFO is "OFF" (100).
```
I'd love there to be a "why" here. Levels between info and off are invalid? I'm going to guess it's because we only expose 
`OFF=100, INFO=20, DEBUGWARNING=15, IO=12, DEBUG=10` from the GUI, and the other values not exposed are `warning=30, error=40, critical=50`

Perhaps a comment like:
```suggestion
	# Don't set a config value outside of what can be set by the GUI
```

feerrenrut commented on source/logHandler.py:
```diff
@@ -350,14 +355,15 @@ def initialize(shouldDoRemoteLogging=False):
 	level = levelNames.get(levelName)
-	if not level or level > log.INFO:
+	# The lone exception to level higher than INFO is "OFF" (100).
+	if not level or log.INFO < level < log.OFF:
```
Given my comment about the values exposed by the GUI, I think the intent (ensure the log value is within a fixed set of values known by the GUI) would be better expressed with:
```suggestion
	if level not in (log.DEBUG, log.IO, log.DEBUGWARNING, log.INFO, log.OFF):
```

feerrenrut commented on source/logHandler.py:
```diff
@@ -313,16 +316,18 @@ def initialize(shouldDoRemoteLogging=False):
 		# Input: kb(desktop):v
 		logFormatter=Formatter("%(levelname)s - %(codepath)s (%(asctime)s.%(msecs)03d):\n%(message)s", "%H:%M:%S")
-		if globalVars.appArgs.secure:
+		if globalVars.appArgs.secure or globalVars.appArgs.noLogging:
```
I think it should be possible to override this for debugging purposes. Perhaps either `-l` `--log-level` or `--debug-logging` could override this.

feerrenrut commented on user_docs/en/userGuide.t2t:
```diff
@@ -2746,12 +2754,13 @@ Following are the command line options for NVDA:
 | -c CONFIGPATH | --config-path=CONFIGPATH | The path where all settings for NVDA are stored |
 | -m | --minimal | No sounds, no interface, no start message, etc. |
-| -s | --secure | Secure mode (disable Python console) |
+| -s | --secure | Secure mode (disables Python console and logging features, used often in secure screens) |
```
In these situations, the admin / IT should have the users temp directory cleared anyway on log out, most applications keep log files. If the user doesn't trust other applications running on the system not to read their log file while they are using the system, then they should not have any illusion of privacy / security on that system. The log file is the least of their worries.

feerrenrut commented:
Happy to take this feature on, since it's already more or less done and gives users  / admin more choices. But I think it's a mistake to think that this is a pathway to privacy or security. For admin trying to ensure that they with GDPR or similar requirements when providing a shared system there are much better ways to ensure the system is cleaned between users. For users who don't trust the system they are using, it should be assumed that every key press, everything displayed, and sound made is recorded. One use case that I think provides an actual benefit, is for uses with limited disk space (though I doubt many users with such limited disk space exist).

josephsl committed:
Merge branch 'master' into i8516noLogging

josephsl committed:
No logging: address review actions. Re #8516.

Changes made:
* log level checks: now a tuple will be checked.
* Allow users to override no logging and logging during secure mode via debug logging or predefined logging levels via command line switches. This means no log will be written if secure mode is active and/or no loggin is specified, and if no debug logging nor custom log levels are given.

josephsl committed:
Launcher: clarify command line options.

Clarified:
* -l (log level): default is set to info, and added options for I/O and debugwarning.
* --debug-warning: add a notice that it can override --no-logging.
* --no-logging: add a notice that this can be overwritten with log levels from command line and if debug logging is on.

josephsl committed:
User guide: edit descriptions for various command line options, especially logging options.

feerrenrut committed:
Update changes file for PR #8596

codeofdusk commented:
The changelog entry for this PR should probably be changed from:

New features: It is possible to turn off NVDA's logging features off by setting logging level to "disabled" from General settings panel. (#8516)

To:

New features: It is possible to turn off NVDA's logging features by setting the logging level to "disabled" from the General settings panel. (#8516)

josephsl commented:
Hi, thanks for confirming it, I’ll queue it for docs19.1 branch (after 2019.1 beta 1 is released). Thanks.

 

From: Bill Dengler <notifications@github.com> 
Sent: Saturday, January 12, 2019 5:32 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Logging: add an option to turn logging off altogether (#8596)

 

The changelog entry for this PR should probably be changed from:

New features: It is possible to turn off NVDA's logging features off by setting logging level to "disabled" from General settings panel. (#8516 <https://github.com/nvaccess/nvda/issues/8516> )

To:

New features: It is possible to turn off NVDA's logging features by setting the logging level to "disabled" from the General settings panel. (#8516 <https://github.com/nvaccess/nvda/issues/8516> )

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8596#issuecomment-453794438> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkH36nTMFN1MxIXHCtaZb91OyXnOIks5vCoyLgaJpZM4Vv0U1> .
